File: src/domain/planner.ts
import type { ContextArm, CreateExperimentRequest, PlannedRun } from "./types";

function hashSeed(value: string): number {
  let hash = 2166136261;
  for (let index = 0; index < value.length; index += 1) {
    hash ^= value.charCodeAt(index);
    hash = Math.imul(hash, 16777619);
  }
  return hash >>> 0;
}

function mulberry32(seed: number): () => number {
  return () => {
    seed |= 0;
    seed = (seed + 0x6d2b79f5) | 0;
    let value = Math.imul(seed ^ (seed >>> 15), 1 | seed);
    value = (value + Math.imul(value ^ (value >>> 7), 61 | value)) ^ value;
    return ((value ^ (value >>> 14)) >>> 0) / 4294967296;
  };
}

function shuffle<T>(items: readonly T[], seed: number): T[] {
  const result = [...items];
  const random = mulberry32(seed);
  for (let index = result.length - 1; index > 0; index -= 1) {
    const target = Math.floor(random() * (index + 1));
    [result[index], result[target]] = [result[target], result[index]];
  }
  return result;
}

export function validateExperimentRequest(request: CreateExperimentRequest): string[] {
  const errors: string[] = [];
  if (!request.name.trim()) errors.push("Experiment name is required.");
  if (!request.dataset.trim()) errors.push("A dataset or manifest is required.");
  if (request.taskIds.length === 0) errors.push("At least one task is required.");
  if (!Number.isInteger(request.repeats) || request.repeats < 1 || request.repeats > 50) {
    errors.push("Repeats must be an integer between 1 and 50.");
  }
  if (!request.arms.includes("none")) errors.push("The none arm is required for a causal baseline.");
  if (request.arms.length < 2) errors.push("At least one context arm is required.");
  if (new Set(request.arms).size !== request.arms.length) errors.push("Context arms must be unique.");
  if (!request.model.provider.trim() || !request.model.model.trim()) {
    errors.push("A frozen provider and model are required.");
  }
  if (!request.agentImage.trim()) errors.push("A pinned container agent image is required.");
  return errors;
}

export function planExperimentRuns(
  experimentId: string,
  request: CreateExperimentRequest,
): PlannedRun[] {
  const errors = validateExperimentRequest(request);
  if (errors.length > 0) throw new Error(errors.join(" "));

  const plans: PlannedRun[] = [];
  for (const taskId of request.taskIds) {
    for (let repeat = 1; repeat <= request.repeats; repeat += 1) {
      const pairId = `${experimentId}:${taskId}:${repeat}`;
      const orderedArms = shuffle<ContextArm>(request.arms, request.seed ^ hashSeed(pairId));
      for (const arm of orderedArms) {
        plans.push({
          id: `${pairId}:${arm}`,
          experimentId,
          pairId,
          taskId,
          repeat,
          arm,
          ordinal: 0,
          status: "queued",
        });
      }
    }
  }

  return shuffle(plans, request.seed).map((plan, ordinal) => ({ ...plan, ordinal }));
}

File: worker/ctxbench_worker/planner.py
from __future__ import annotations

import hashlib
import random

from .models import ExperimentSpec, PlannedRun
from .workflows import normalize_workflow


def validate_experiment(spec: ExperimentSpec) -> list[str]:
    errors: list[str] = []
    if not spec.name.strip():
        errors.append("Experiment name is required.")
    if not spec.dataset.strip():
        errors.append("A dataset or manifest is required.")
    if not spec.task_ids:
        errors.append("At least one task is required.")
    if len(set(spec.task_ids)) != len(spec.task_ids):
        errors.append("Task IDs must be unique.")
    if not 1 <= spec.repeats <= 50:
        errors.append("Repeats must be between 1 and 50.")
    if "none" not in spec.arms:
        errors.append("The none arm is required for a causal baseline.")
    if len(spec.arms) < 2:
        errors.append("At least one context arm is required.")
    if len(set(spec.arms)) != len(spec.arms):
        errors.append("Context arms must be unique.")
    if not spec.model.provider.strip() or not spec.model.model.strip():
        errors.append("Provider and model must be frozen before planning.")
    if not spec.agent_image.strip():
        errors.append("A pinned container agent image is required.")
    for workflow in (spec.builder_workflow, spec.solver_workflow):
        try:
            normalize_workflow(workflow)
        except ValueError as error:
            errors.append(str(error))
    return errors


def _derived_seed(seed: int, value: str) -> int:
    digest = hashlib.sha256(f"{seed}:{value}".encode()).digest()
    return int.from_bytes(digest[:8], "big")


def plan_runs(experiment_id: str, spec: ExperimentSpec) -> list[PlannedRun]:
    errors = validate_experiment(spec)
    if errors:
        raise ValueError(" ".join(errors))

    result: list[PlannedRun] = []
    for task_id in spec.task_ids:
        for repeat in range(1, spec.repeats + 1):
            pair_id = f"{experiment_id}:{task_id}:{repeat}"
            arms = list(spec.arms)
            random.Random(_derived_seed(spec.seed, pair_id)).shuffle(arms)
            for arm in arms:
                result.append(
                    PlannedRun(
                        id=f"{pair_id}:{arm}",
                        experiment_id=experiment_id,
                        pair_id=pair_id,
                        task_id=task_id,
                        repeat=repeat,
                        arm=arm,
                        ordinal=0,
                    )
                )

    random.Random(spec.seed).shuffle(result)
    return [PlannedRun(**{**run.__dict__, "ordinal": ordinal}) for ordinal, run in enumerate(result)]
